### PR TITLE
storage: Introduce isDefinitelyEmpty API

### PR DESCRIPTION
**Pull request by breezewish**
### What problem does this PR solve?

Issue Number: ref #5237

Problem Summary:

### What is changed and how it works?

- Added a new API: `Segment::isDefinitelyEmpty`.

  It returns true when the snapshot is definitely empty.
  It returns false when the snapshot is not definitely empty, e.g. not empty, or may be empty in some rare cases.

  This API is used to determine whether it is safe to perform replaceData for the segment when we trying to use replaceData for the ingest SST process. It could be also helpful for https://github.com/pingcap/tiflash/issues/5816

  This API does not respond an accurate isEmpty result in certain cases, in order to minimize the cost.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

**Review by lidezhu — APPROVED** [review]
LGTM

File: dbms/src/Storages/DeltaMerge/Segment.cpp [hunk 1]
```diff
@@ -373,6 +375,77 @@ bool Segment::write(DMContext & dm_context, const RowKeyRange & delete_range)
...
+    auto columns_to_read = std::make_shared<ColumnDefines>();
+    columns_to_read->push_back(getExtraHandleColumnDefine(is_common_handle));
+    auto read_ranges = RowKeyRanges{rowkey_range};
+
+    SkippableBlockInputStreams streams;
+    for (const auto & file : segment_snap->stable->getDMFiles())
+    {
+        DMFileBlockInputStreamBuilder builder(dm_context.db_context);
+        auto stream = builder
+                          .setRowsThreshold(std::numeric_limits<UInt64>::max()) // TODO: May be we could have some better settings
```
Comment (JaySon-Huang) [inline dbms/src/Storages/DeltaMerge/Segment.cpp:401]: I think we don't need this line as we only try to read one row from the stream

File: dbms/src/Storages/DeltaMerge/Segment.cpp [hunk 2]
```diff
@@ -373,6 +375,77 @@ bool Segment::write(DMContext & dm_context, const RowKeyRange & delete_range)
...
+        ProfileEvents::increment(ProfileEvents::DMSegmentIsEmptyFastPath);
+        return true;
+    }
+
+    ProfileEvents::increment(ProfileEvents::DMSegmentIsEmptySlowPath);
+
+    // Build read stream
+    auto columns_to_read = std::make_shared<ColumnDefines>();
+    columns_to_read->push_back(getExtraHandleColumnDefine(is_common_handle));
+    auto read_ranges = RowKeyRanges{rowkey_range};
```
Comment (JaySon-Huang) [inline dbms/src/Storages/DeltaMerge/Segment.cpp:394]: what about using `getInputStreamModeFast`?

File: dbms/src/Storages/DeltaMerge/Segment.cpp (hunk 1 above)
Comment (breezewish) [reply dbms/src/Storages/DeltaMerge/Segment.cpp:401]: We may also read multiple rows in some rare cases, for example, the whole block may be filtered out by SegmentRange.

File: dbms/src/Storages/DeltaMerge/Segment.cpp (hunk 2 above)
Comment (breezewish) [reply dbms/src/Storages/DeltaMerge/Segment.cpp:394]: I'm not sure. My intuition was to not rely on fast mode, as fast mode may change behavior at any time. Currently fast mode will additionally do filter for del_mark, while we don't need it as well.

Comment (JaySon-Huang) [reply dbms/src/Storages/DeltaMerge/Segment.cpp:394]: OK

File: dbms/src/Storages/DeltaMerge/Segment.cpp [hunk 3]
```diff
@@ -373,6 +375,77 @@ bool Segment::write(DMContext & dm_context, const RowKeyRange & delete_range)
...
+    {
+        DMFileBlockInputStreamBuilder builder(dm_context.db_context);
+        auto stream = builder
+                          .setRowsThreshold(std::numeric_limits<UInt64>::max()) // TODO: May be we could have some better settings
+                          .onlyReadOnePackEveryTime()
+                          .build(file, *columns_to_read, read_ranges);
+        streams.push_back(stream);
+    }
+
+    BlockInputStreamPtr stable_stream = std::make_shared<ConcatSkippableBlockInputStream>(streams);
```
Comment (JaySon-Huang) [inline dbms/src/Storages/DeltaMerge/Segment.cpp:407]: We can only build the stable_stream after there is no data read from `delta_stream`. By doing this we can save some unnecessary IO for min-max index.
But now it is also fine.

**Review by JaySon-Huang — APPROVED** [review]
LGTM

File: dbms/src/Storages/DeltaMerge/Segment.cpp (hunk 3 above)
Comment (breezewish) [reply dbms/src/Storages/DeltaMerge/Segment.cpp:407]: Good catch!

**Pull request merged**